### PR TITLE
DOCSUP-7551: Documented nested with multiple nesting and subcolumns of complex types

### PR DESCRIPTION
**Pull request by gyuton**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Documentation for #17310

### CODE REVIEW

File: docs/en/sql-reference/data-types/array.md [hunk 1]
```diff
@@ -74,4 +74,28 @@ Received exception from server (version 1.1.54388):
...
+It is possible to find the size of an array by using the `size0` subcolumn without reading the whole column. For multi-dimensional arrays you can use size`n-1`, where `n` is the wanted dimension.
+
+Query:
+
+```sql
+CREATE TABLE t_arr (a Array(UInt32)) ENGINE = MergeTree ORDER BY tuple();
+
+CREATE TABLE t_arr (`arr` Array(Array(Array(UInt32)))) ENGINE = MergeTree ORDER BY tuple();
+
+insert into t_arr values ([[[12, 13, 0, 1],[12]]]);
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/array.md:88]:
```suggestion
INSERT INTO t_arr VALUES ([[[12, 13, 0, 1],[12]]]);
```

File: docs/en/sql-reference/data-types/array.md [hunk 2]
```diff
@@ -74,4 +74,28 @@ Received exception from server (version 1.1.54388):
...
 ```
 
+## Array Size {#array-size}
+
+It is possible to find the size of an array by using the `size0` subcolumn without reading the whole column. For multi-dimensional arrays you can use size`n-1`, where `n` is the wanted dimension.
+
+Query:
+
+```sql
+CREATE TABLE t_arr (a Array(UInt32)) ENGINE = MergeTree ORDER BY tuple();
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/array.md:84]:
```suggestion

```

File: docs/en/sql-reference/data-types/array.md [hunk 3]
```diff
@@ -74,4 +74,28 @@ Received exception from server (version 1.1.54388):
...
+
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/array.md:85]:
```suggestion
```

File: docs/en/sql-reference/data-types/tuple.md [hunk 4]
```diff
@@ -47,4 +47,32 @@ SELECT tuple(1, NULL) AS x, toTypeName(x)
 └──────────┴─────────────────────────────────┘
 ```
 
+## Addressing Tuple Element {#addressing-tuple-element}
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/tuple.md:50]:
```suggestion
## Addressing Tuple Elements {#addressing-tuple-elements}
```

File: docs/en/sql-reference/data-types/nullable.md [hunk 5]
```diff
@@ -20,6 +20,31 @@ To store `Nullable` type values in a table column, ClickHouse uses a separate fi
 !!! info "Note"
     Using `Nullable` almost always negatively affects performance, keep this in mind when designing your databases.
 
+## Finding NULL {#finding-null}
+
+It is possible to find `NULL` values in a column by using `null` subcolumn without reading the whole column. It returns `1` if the corresponding value is `NULL` and `0` otherwise.
+
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/nullable.md:26]:
```suggestion
**Example**
```

File: docs/ru/sql-reference/data-types/array.md [hunk 6]
```diff
@@ -5,9 +5,7 @@ toc_title: Array(T)
 
 # Array(T) {#data-type-array}
 
-Массив из элементов типа `T`.
-
-`T` может любым, в том числе, массивом. Таким образом поддержаны многомерные массивы.
+Массив из элементов типа `T`. `T` может любым, в том числе, массивом. Таким образом поддержаны многомерные массивы.
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/array.md:8]:
```suggestion
Массив из элементов типа `T`. `T` может любым, в том числе массивом. Таким образом поддерживаются многомерные массивы.
```

File: docs/ru/sql-reference/data-types/array.md [hunk 7]
```diff
@@ -5,9 +5,7 @@ toc_title: Array(T)
...
 
 ## Создание массива {#sozdanie-massiva}
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/array.md:10]:
```suggestion
## Создание массива {#creating-an-array}
```

File: docs/ru/sql-reference/data-types/array.md [hunk 8]
```diff
@@ -76,3 +74,24 @@ Received exception from server (version 1.1.54388):
 Code: 386. DB::Exception: Received from localhost:9000, 127.0.0.1. DB::Exception: There is no supertype for types UInt8, String because some of them are String/FixedString and some of them are not.
 ```
 
+## Размер массива {#array-size}
+
+Узнать размер массива можно с помощью подстолбца `size0` без чтения всего столбца. Для многомерных массивов можно использовать подстолбец `sizen-1`, где `n` — требуемое измерение.
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/array.md:79]:
```suggestion
Узнать размер массива можно с помощью подстолбца `size0` без чтения всего столбца. Для многомерных массивов можно использовать подстолбец `sizeN-1`, где `N` — требуемое измерение.
```

File: docs/en/sql-reference/data-types/array.md [hunk 9]
```diff
@@ -74,4 +74,24 @@ Received exception from server (version 1.1.54388):
 Code: 386. DB::Exception: Received from localhost:9000, 127.0.0.1. DB::Exception: There is no supertype for types UInt8, String because some of them are String/FixedString and some of them are not.
 ```
 
-[Original article](https://clickhouse.tech/docs/en/data_types/array/) <!--hide-->
+## Array Size {#array-size}
+
+It is possible to find the size of an array by using the `size0` subcolumn without reading the whole column. For multi-dimensional arrays you can use `sizen-1`, where `n` is the wanted dimension.
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/array.md:79]:
```suggestion
It is possible to find the size of an array by using the `size0` subcolumn without reading the whole column. For multi-dimensional arrays you can use `sizeN-1`, where `N` is the wanted dimension.
```

File: docs/ru/sql-reference/data-types/array.md [hunk 10]
```diff
@@ -76,3 +74,24 @@ Received exception from server (version 1.1.54388):
...
+
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/array.md:80]:
```suggestion
**Пример**
```

File: docs/ru/sql-reference/data-types/nullable.md [hunk 11]
```diff
@@ -27,6 +27,31 @@ toc_title: Nullable
 !!! info "Info"
     Почти всегда использование `Nullable` снижает производительность, учитывайте это при проектировании своих баз.
 
+## Поиск NULL {#finding-null}
+
+Найти в столбце значения `NULL` можно с помощью подстобца `null`, при этом чтение всего столбца не происходит. Подстолбец возвращает `1`, если соответственное значение равно `NULL`, и `0` в противоположном случае.
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/nullable.md:32]:
```suggestion
Найти в столбце значения `NULL` можно с помощью подстолбца `null`, при этом чтение всего столбца не происходит. Подстолбец возвращает `1`, если соответственное значение равно `NULL`, и `0` в противоположном случае.
```

File: docs/ru/sql-reference/data-types/nullable.md [hunk 12]
```diff
@@ -27,6 +27,31 @@ toc_title: Nullable
...
+
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/nullable.md:33]:
```suggestion
**Пример**
```

File: docs/en/sql-reference/data-types/tuple.md [hunk 13]
```diff
@@ -47,4 +47,32 @@ SELECT tuple(1, NULL) AS x, toTypeName(x)
 └──────────┴─────────────────────────────────┘
 ```
 
+## Addressing Tuple Elements {#addressing-tuple-elements}
+
+It is possible to read elements of named tuples using indexes and names:
+
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/tuple.md:53]:
```suggestion
**Example**
```

File: docs/ru/sql-reference/data-types/nullable.md [hunk 14]
```diff
@@ -27,6 +27,31 @@ toc_title: Nullable
...
+``` text
+┌─n.null─┐
+│      0 │
+│      1 │
+│      0 │
+│      1 │
+└────────┘
+```
+
 ## Пример использования {#primer-ispolzovaniia}
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/nullable.md:55]:
```suggestion
## Пример использования {#usage-example}
```

File: docs/ru/sql-reference/data-types/tuple.md [hunk 15]
```diff
@@ -47,3 +47,30 @@ SELECT tuple(1,NULL) AS x, toTypeName(x)
 └──────────┴─────────────────────────────────┘
 ```
 
+## Адресация элементов кортежа {#addressing-tuple-elements}
+
+Можно считывать элементы кортежей с помощью индексов и имен:
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/tuple.md:52]:
```suggestion
К элементам кортежа можно обращаться по индексу и по имени:
```

**Comment by gyuton** [comment]
@CurtizJ Hi! Can you review this, please?

File: docs/en/sql-reference/data-types/nested-data-structures/nested.md [hunk 16]
```diff
@@ -34,7 +34,7 @@ CREATE TABLE test.visits
 
 This example declares the `Goals` nested data structure, which contains data about conversions (goals reached). Each row in the ‘visits’ table can correspond to zero or any number of conversions.
 
-Only a single nesting level is supported. Columns of nested structures containing arrays are equivalent to multidimensional arrays, so they have limited support (there is no support for storing these columns in tables with the MergeTree engine).
+Arbitrary levels of nesting are supported. Columns of nested structures containing arrays are equivalent to multidimensional arrays, so they have limited support (there is no support for storing these columns in tables with the MergeTree engine).
```
Comment (CurtizJ) [inline docs/en/sql-reference/data-types/nested-data-structures/nested.md:37]: Need to mention, that it's controlled by setting `flatten_nested` (`true` by default). Arbitrary levels of nesting are supported only, when `flatten_nested` is false. Also need to mention, that format of data, inserted to columns of type `Nested` changes, when `flatten_nested` is false (separate arrays vs array of tuples).

Example:

```sql
SET flatten_nested = 0;

CREATE TABLE t_nest(n Nested(a UInt32, b UInt32)) ENGINE = MergeTree ORDER BY tuple();

SHOW CREATE TABLE t_nest;

┌─statement───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE default.t_nest
(
    `n.a` Array(UInt32),
    `n.b` Array(UInt32)
)
ENGINE = MergeTree
ORDER BY tuple()
SETTINGS index_granularity = 8192 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

INSERT INTO t_nest VALUES ([1, 2], [3, 4]);
```

```sql
SET flatten_nested = 0;

CREATE TABLE t_nest(n Nested(a UInt32, b UInt32)) ENGINE = MergeTree ORDER BY tuple();

SHOW CREATE TABLE t_nest;

┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE default.t_nest
(
    `n` Nested(a UInt32, b UInt32)
)
ENGINE = MergeTree
ORDER BY tuple()
SETTINGS index_granularity = 8192 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

INSERT INTO t_nest VALUES ([(1, 3), (2, 4)])
```

File: docs/en/operations/settings/settings.md [hunk 17]
```diff
@@ -3065,4 +3065,66 @@ SELECT
 FROM fuse_tbl
 ```
 
+## flatten_nested {#flatten-nested}
+
+Sets the data format of a [nested](../../sql-reference/data-types/nested-data-structures/nested.md) columns.
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3070]:
```suggestion
Sets the data format of [nested](../../sql-reference/data-types/nested-data-structures/nested.md) columns.
```

File: docs/en/operations/settings/settings.md [hunk 18]
```diff
@@ -3065,4 +3065,66 @@ SELECT
...
+-   0 — Nested column is flattened to separate arrays.
+-   1 — Nested column is flattened to a singe array of tuples.
+
+Default value: `1`.
+
+**Usage**
+
+If the setting is set to `0`, it is possible to use an arbitrary level of nesting.
+
+With the setting set to `1`:
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3083]: **Example**

File: docs/en/operations/settings/settings.md [hunk 19]
```diff
@@ -3065,4 +3065,66 @@ SELECT
...
+
+``` sql
+CREATE TABLE t_nest (`n` Nested(a UInt32, b UInt32)) ENGINE = MergeTree ORDER BY tuple();
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3086]:
```suggestion
CREATE TABLE t_nest (`n` Nested(a UInt32, b UInt32)) ENGINE = MergeTree ORDER BY tuple() SETTINGS flatten_nested = 1;
```

File: docs/en/operations/settings/settings.md [hunk 20]
```diff
@@ -3065,4 +3065,66 @@ SELECT
...
+    `n.a` Array(UInt32),
+    `n.b` Array(UInt32)
+)
+ENGINE = MergeTree
+ORDER BY tuple()
+SETTINGS index_granularity = 8192 │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+```
+
+With the setting set to `0`:
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3106]:
```suggestion
```

File: docs/en/operations/settings/settings.md (hunk 18 above)
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3083]:
```suggestion
**Example**

Query:
```

File: docs/en/sql-reference/data-types/nested-data-structures/nested.md [hunk 21]
```diff
@@ -34,7 +34,7 @@ CREATE TABLE test.visits
 
 This example declares the `Goals` nested data structure, which contains data about conversions (goals reached). Each row in the ‘visits’ table can correspond to zero or any number of conversions.
 
-Only a single nesting level is supported. Columns of nested structures containing arrays are equivalent to multidimensional arrays, so they have limited support (there is no support for storing these columns in tables with the MergeTree engine).
+Arbitrary levels of nesting are supported, but to use more than one level you need to set the setting [flatten_nested](../../../operations/settings/settings.md#flatten-nested) to `0`. Columns of nested structures containing arrays are equivalent to multidimensional arrays, so they have limited support (there is no support for storing these columns in tables with the MergeTree engine).
```
Comment (olgarev) [inline docs/en/sql-reference/data-types/nested-data-structures/nested.md:37]:
```suggestion
When [flatten_nested](../../../operations/settings/settings.md#flatten-nested) is set to `0` (which is not by default), arbitrary levels of nesting are supported. In this case columns of nested structures, containing arrays, are flattened to multidimensional arrays, so they have limited support (there is no support for storing these columns in tables with the [MergeTree](../../engines/table-engines/mergetree-family/mergetree.md) engine).
```

File: docs/ru/operations/settings/settings.md [hunk 22]
```diff
@@ -2956,4 +2956,66 @@ SELECT
...
+-   0 — вложенный столбец преобразуется к отдельным массивам.
+-   1 — вложенный столбец преобразуется к массиву кортежей.
+
+Значение по умолчанию: `1`.
+
+**Использование**
+
+Если установлено значение `0`, возможно использовать любой уровень вложенности.
+
+С значением настройки `1`:
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:2974]:
```suggestion
Если установлено значение `0`, можно использовать любой уровень вложенности.

**Примеры**

Запрос:
```

File: docs/ru/operations/settings/settings.md [hunk 23]
```diff
@@ -2956,4 +2956,66 @@ SELECT
...
+
+``` sql
+CREATE TABLE t_nest (`n` Nested(a UInt32, b UInt32)) ENGINE = MergeTree ORDER BY tuple();
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:2977]:
```suggestion
SET flatten_nested = 1;

CREATE TABLE t_nest (`n` Nested(a UInt32, b UInt32)) ENGINE = MergeTree ORDER BY tuple();
```

File: docs/ru/operations/settings/settings.md [hunk 24]
```diff
@@ -2956,4 +2956,66 @@ SELECT
...
+    `n.a` Array(UInt32),
+    `n.b` Array(UInt32)
+)
+ENGINE = MergeTree
+ORDER BY tuple()
+SETTINGS index_granularity = 8192 │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+```
+
+С значением настройки `0`:
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:2997]:
```suggestion
Запрос:
```

File: docs/ru/sql-reference/data-types/nested-data-structures/nested.md [hunk 25]
```diff
@@ -29,7 +29,7 @@ CREATE TABLE test.visits
 
 В этом примере объявлена вложенная структура данных `Goals`, содержащая данные о достижении целей. Каждой строке таблицы visits может соответствовать от нуля до произвольного количества достижений целей.
 
-Поддерживается только один уровень вложенности. Столбцы вложенных структур, содержащие массивы, эквивалентны многомерным массивам, поэтому их поддержка ограничена (не поддерживается хранение таких столбцов в таблицах с движком семейства MergeTree).
+Поддерживаются любые уровни вложенности, но чтобы использовать уровень больше одного, необходимо установить `0` настройке [flatten_nested](../../../operations/settings/settings.md#flatten-nested). Столбцы вложенных структур, содержащие массивы, эквивалентны многомерным массивам, поэтому их поддержка ограничена (не поддерживается хранение таких столбцов в таблицах с движком семейства MergeTree).
```
Comment (olgarev) [inline docs/ru/sql-reference/data-types/nested-data-structures/nested.md:32]:
```suggestion
Если настройка [flatten_nested](../../../operations/settings/settings.md#flatten-nested) установлена в значение `0`, поддерживаются любые уровни вложенности. В этом случае столбцы вложенных структур, содержащие массивы, преобразуются к многомерным массивам, поэтому их поддержка ограничена (хранение таких столбцов в таблицах с движком семейства [MergeTree](../../engines/table-engines/mergetree-family/mergetree.md) не поддерживается).
```

File: docs/en/operations/settings/settings.md [hunk 26]
```diff
@@ -3065,4 +3065,69 @@ SELECT
...
 ```
 
+## flatten_nested {#flatten-nested}
+
+Sets the data format of a [nested](../../sql-reference/data-types/nested-data-structures/nested.md) columns.
+
+Possible values:
+
+-   0 — Nested column is flattened to separate arrays.
+-   1 — Nested column is flattened to a singe array of tuples.
```
Comment (CurtizJ) [inline docs/en/operations/settings/settings.md:3075]: So, it's not flattened in that case.

Comment (CurtizJ) [reply docs/en/operations/settings/settings.md:3075]: And nested column is flattened to arrays, when `flatten_nested = 1`.

**Review by CurtizJ — APPROVED** [review]

**Pull request merged**